packages/vite/src/node/plugins/define.ts

```diff
@@ -116,14 +129,30 @@ export function definePlugin(config: ResolvedConfig): Plugin {
         return
       }
 
-      const [define, pattern] = ssr ? ssrPattern : defaultPattern
+      const [define, pattern, undefinedPattern, banner] = ssr
+        ? ssrPattern
+        : defaultPattern
       if (!pattern) return
 
       // Check if our code needs any replacements before running esbuild
       pattern.lastIndex = 0
       if (!pattern.test(code)) return
 
-      return await replaceDefine(code, id, define, config)
+      // process undefined imports.meta.env properties
+      const defineWithUndefined = { ...define }
+      undefinedPattern.lastIndex = 0
+      for (const undefinedEnvKey of [...code.matchAll(undefinedPattern)]) {
+        defineWithUndefined[undefinedEnvKey[0]] = 'undefined'
+      }
```

Comment:
Instead of dynamically updating the `define` list, could we do a post replacement instead after `replaceDefine` is called? That way this part can be more performant and not constantly creating new objects.

With a post-replacement, I think we should also simplify the regex as `/__vite_import_meta_env__.(.*)\b/` instead of constructing the regex like above. We can then compare the captured string if it's part of the keys of `import.meta.env`. In the past, we found that a giant regex like that is quite slow.